File: DFS, BFS/2667.py
#단지번호붙이기
def dfs(x, y):
    global cnt
    if x<0 or x>=n or y<0 or y>=n:
        return False
    if graph[x][y] == 1:
        cnt += 1
        graph[x][y] = 0
        for i in range(4):
            dfs(x+dx[i],y+dy[i])
        return True
    return False

graph = []
dx = [-1,1,0,0]
dy = [0,0,-1,1]
result = []
cnt = 0
n = int(input())
for _ in range(n):
    graph.append(list(map(int, input())))
for i in range(n):
    for j in range(n):
        if dfs(i,j) == True:
            result.append(cnt)
            cnt = 0
result.sort()
print(len(result))
for i in result:
    print(i)

# bfs 풀이
# from collections import deque
# def bfs(x,y):
#     queue = deque()
#     graph[x][y] = 0
#     cnt = 1
#     queue.append((x,y))
#     while queue:
#         x, y = queue.popleft()
#         for i in range(4):
#             nx = x + dx[i]
#             ny = y + dy[i]
#             if 0<=nx<n and 0<=ny<n and graph[nx][ny] == 1:
#                 queue.append((nx,ny))
#                 graph[nx][ny] = 0
#                 cnt += 1
#     return cnt

# graph = []
# dx = [-1,1,0,0]
# dy = [0,0,-1,1]
# result = []
# n = int(input())
# for _ in range(n):
#     graph.append(list(map(int, input())))
# for i in range(n):
#     for j in range(n):
#         if graph[i][j] == 1:
#             result.append(bfs(i, j))
# result.sort()
# print(len(result))
# for i in result:
#     print(i)

File: DFS, BFS/2667.js
// 단지번호붙이기
// let input = require("fs")
//   .readFileSync("/dev/stdin")
//   .toString()
//   .trim()
//   .split("\n");
let input = `7
0110100
0110101
1110101
0000111
0100000
0111110
0111000`.split("\n");

let n = Number(input[0]);
let graph = [];
for (let i = 1; i < input.length; i++) {
  let temp = input[i].split("").map((v) => Number(v));
  graph.push(temp);
}

let total = 0;
let results = [];
let homes = 1;
let dx = [0, 1, 0, -1];
let dy = [1, 0, -1, 0];

for (let i = 0; i < n; i++) {
  for (let j = 0; j < n; j++) {
    if (graph[i][j]) {
      total++;
      graph[i][j] = 0;
      //   dfs(i, j);
      bfs(i, j);
      results.push(homes);
      homes = 1;
    }
  }
}
results.sort((a, b) => a - b);
console.log(total + "\n" + results.join("\n"));

function dfs(x, y) {
  for (let i = 0; i < 4; i++) {
    let nx = x + dx[i];
    let ny = y + dy[i];
    if (0 <= nx && nx < n && 0 <= ny && ny < n && graph[nx][ny]) {
      graph[nx][ny] = 0;
      homes++;
      dfs(nx, ny);
    }
  }
}

function bfs(x, y) {
  let q = [];
  q.push([x, y]);
  while (q.length) {
    let [x, y] = q.shift();
    for (let i = 0; i < 4; i++) {
      let nx = x + dx[i];
      let ny = y + dy[i];
      if (0 <= nx && nx < n && 0 <= ny && ny < n && graph[nx][ny]) {
        graph[nx][ny] = 0;
        homes++;
        q.push([nx, ny]);
      }
    }
  }
}
